File: 2501-longest-square-streak-in-an-array/2501-longest-square-streak-in-an-array.java
class Solution {
    public int longestSquareStreak(int[] nums) {
        Arrays.sort(nums);
        int ans = 0;
        var dp = new HashMap<Integer, Integer>();
        for(int num : nums) {
            int sqr = (int) Math.sqrt(num);
            if(sqr * sqr == num)
                dp.put(num, 1 + dp.getOrDefault(sqr, 0));
            else dp.put(num, 1);
            ans = Math.max(ans, dp.get(num));
        }
        return ans > 1 ? ans : -1;
    }
}

File: 2501-longest-square-streak-in-an-array/2501-longest-square-streak-in-an-array.cpp
class Solution {
public:
    int longestSquareStreak(vector<int>& nums) {
        // Solving this DP problem in 2 steps:
        // Step 1: sort the nums array
        // Step 2: for each element in nums, check if it is a perfect square
        // if it is, increase the count of elements inside our current streak
        // dp[num] = 1 + dp[squareRoot]
        // else just mark it as one so that it can be used later as a squareRoot for
        // another larger number
        sort(begin(nums), end(nums));
        unordered_map<int, int> dp;
        int ans = 0;
        for(int &num : nums) {
            int sqr = sqrt(num);
            if(sqr * sqr == num)
                dp[num] = 1 + dp[sqr];
            else
                dp[num] = 1;
            ans = max(ans, dp[num]);
        }
        return ans > 1 ? ans : -1;
    }
};